Core/include/Acts/TrackFitting/GainMatrixUpdater.hpp

```diff
@@ -11,44 +11,32 @@
 #include "Acts/EventData/Measurement.hpp"
 #include "Acts/EventData/MeasurementHelpers.hpp"
 #include "Acts/EventData/MultiTrajectory.hpp"
-#include "Acts/EventData/TrackParameters.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
 #include "Acts/TrackFitting/KalmanFitterError.hpp"
-#include "Acts/TrackFitting/detail/VoidKalmanComponents.hpp"
-#include "Acts/Utilities/Definitions.hpp"
 #include "Acts/Utilities/Logger.hpp"
 #include "Acts/Utilities/Result.hpp"
 
-#include <memory>
-#include <variant>
-
 namespace Acts {
 
-/// @brief Update step of Kalman Filter using gain matrix formalism
+/// Kalman update step using the gain matrix formalism.
 class GainMatrixUpdater {
  public:
-  /// @brief Public call operator for the boost visitor pattern
-  ///
-  /// @tparam track_state_t Type of the track state for the update
-  ///
-  /// @param gctx The current geometry context object, e.g. alignment
-  /// @param trackState the measured track state
-  /// @param direction the navigation direction
+  /// Run the Kalman update step for a single trajectory state.
   ///
-  /// @return Bool indicating whether this update was 'successful'
-  /// @note Non-'successful' updates could be holes or outliers,
-  ///       which need to be treated differently in calling code.
-  template <typename track_state_t>
-  Result<void> operator()(const GeometryContext& /*gctx*/,
-                          track_state_t trackState,
-                          const NavigationDirection& direction = forward,
-                          LoggerWrapper logger = getDummyLogger()) const {
+  /// @tparam source_link_t The type of source link
+  /// @tparam kMeasurementSizeMax
+  /// @param[in] gctx The current geometry context object, e.g. alignment
+  /// @param[in,out] trackState The track state
+  /// @param[in] direction The navigation direction
+  /// @param[in] logger Where to write logging information to
+  template <typename source_link_t, size_t kMeasurementSizeMax>
+  Result<void> operator()(
+      const GeometryContext& /*gctx*/,
+      detail_lt::TrackStateProxy<source_link_t, kMeasurementSizeMax, false>&
```

Comment:
Let's use `typename MultiTrajectory<source_link_t>::TrackStateProxy` so we don't need to name `detail_lt`.

Reply:
That was my initial though, too. But that does not seem to work for reasons unclear to me. I had to use the implementation type in `detail_lt` to make it work. I agree that it is not the nicest thing to do; while the implementation type exists in the `detail_lt` namespace it is being exported into the public interface anyways and so I think its Ok to take it as input here.

Reply:
I see. I think I ran into the same issue when writing this initially, and that's why I chose to use a blanket template with static assertions. I still prefer that over having `detail_lt` in the argument typelist, but ok.